src/System.Management.Automation/namespaces/FileSystemProvider.cs

```diff
@@ -31,6 +31,56 @@
 
 namespace Microsoft.PowerShell.Commands
 {
+    #region InodeTracker
+    /// <summary>
+    /// Tracks visited files/directories by caching their device IDs and inodes.
+    /// </summary>
+    internal class InodeTracker
+    {
+        private Dictionary<(UInt64, UInt64), bool>  _visitations;
+
+        /// <summary>
+        /// Construct a new InodeTracker with an initial path
+        /// </summary>
+        internal InodeTracker(string path)
+        {
+            _visitations = new Dictionary<(UInt64, UInt64), bool>();
+            var inodeData = (0UL, 0UL);
+
+            if (InternalSymbolicLinkLinkCodeMethods.GetInodeData(path, out inodeData))
+            {
+                _visitations[inodeData] = true;
+            }
+        }
+
+        /// <summary>
+        /// Try to mark a path as having been visited.
+        /// </summary>
+        /// <param name="path">Path to the file or directory to be marked as visited.</param>
+        /// <returns>
+        /// False if the path has already been visited.
+        /// True if the path has not yet been visited.
+        /// </returns>
+        /// <remarks>
+        /// If the path has not yet been visited then before this method returns it marks
+        /// the path as having been visited.
+        /// </remarks>
+        internal bool TryVisitPath(string path)
```

Comment:
This method goes against the convention of `Try<Verb>` pattern. Take `TryAdd` as an instance, if added, then return true; if not added, then return false. Following that conversion, `TryVisitPath` should be: if visited, then return true; if not visited, then return false. So `TryVisitPath` makes it confusing understanding the code.

How about making it `IsPathVisited(bool visitIfNotAlready = false)`?
```CSharp
internal bool IsPathVisited(bool visitIfNotAlready = false)
{
    bool isPathVisited = false;
    if (InternalSymbolicLinkLinkCodeMethods.GetInodeData(path, out (UInt64, UInt64) inodeData))
    {
        if (visitIfNotAlready)
        {
            isPathVisited = !_visitations.Add(inodeData)
        }
        else
        {
            isPathVisited = _visitations.Contains(inodeData)
        }
    }
    return isPathVisited;
}
```